nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,452 @@
+        )
+
+        self._loss_function_unsupervised = self._loss_function = config.get(
+            "loss_function_unsupervised", config.get("loss_function", F.mse_loss)
+        )
+        self._loss_function_supervised = config.get(
+            "loss_function_supervised", torch.nn.functional.cross_entropy
+        )
+
+        self.total_batches_unsupervised = self.total_batches
+        self.total_batches_supervised = len(self.supervised_loader)
+        self.batches_in_epoch_supervised = config.get(
+            "batches_in_epoch_supervised", sys.maxsize
+        )
+
+        self.lr_scheduler_classifier = self.create_lr_scheduler_classifier(
+            config, self.classifier_optimizer, self.total_batches_supervised
+        )
+        self.step_lr_every_batch_classifier = config.get(
+            "lr_scheduler_classifier_step_every_batch", False
+        )
+        if isinstance(self.lr_scheduler_classifier, (OneCycleLR, ComposedLRScheduler)):
+            self.step_lr_every_batch_classifier = True
+
+        self.reset_classifier_on_validate = config.get(
+            "reset_classifier_on_validate", False
+        )
+
+        self.supervised_training_epochs_per_validation = config.get(
+            "supervised_training_epochs_per_validation", 3
+        )
+
+    @classmethod
+    def create_classifier(cls, config, device):
+        """
+        Create `torch.nn.Module` model from an experiment config
+        :param config:
+            - classifier_model_class: Model class. Must inherit from "torch.nn.Module"
+            - classifier_model_args: model model class arguments passed to the
+                                     constructor
+            - classifier_init_batch_norm: Whether or not to Initialize running batch
+                                          norm mean to 0.
+            - classifier_checkpoint_file: if not None, will start from this model. The
+                                          model must have the same model_args and
+                                          model_class as the current experiment.
+            - classifier_load_checkpoint_args_classifier: args to be passed to
+                                               `load_state_from_checkpoint`
+        :param device:
+            Pytorch device
+        :return:
+                Model instance
+        """
+        return create_model(
+            model_class=config["classifier_model_class"],
+            model_args=config.get("classifier_model_args", {}),
+            init_batch_norm=config.get("classifier_init_batch_norm", False),
+            device=device,
+            checkpoint_file=config.get("classifier_checkpoint_file", None),
+            load_checkpoint_args=config.get(
+                "classifier_load_checkpoint_args_classifier", {}
+            ),
+        )
+
+    @classmethod
+    def create_classifier_optimizer(cls, config, model):
+        """
+        Create optimizer for the classifier from an experiment config.
```

Comment:
Defining new class methods for creating the classifier, it's optimizer, LR scheduler, etc is one way to go, but another method that I'd strongly recommend considering is passing in a 'pseudo config' where the value of `optimizer_class` and `optimizer_args` simply point to those of `classifier_optimizer_class` and `classifier_optimizer_args` in the original dict. E.g., it would look like

```
new_config = deepcopy(config)
new_config.update(
    optimizer_class=config.get("classifier_optimizer_class"),
    optimizer_args=config.get("classifier_optimizer_args")
)

self.classifier_optimizer = self.create_optimizer(new_config)
```

and this same idea can be used to create the classifier model itself, etc. This way, you don't need to define these new methods which are almost identical to the ones in `SupervisedExperiment`.

Reply:
Great suggestion. I have taken this a step further to change the config passed in to require a "classifier_config" dictionary which has the keys of a regular config, but is specific to the classifier. That way, you can define a single classifier config and use it in several different places.

Reply:
Nice!